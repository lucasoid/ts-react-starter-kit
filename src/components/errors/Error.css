.errorMessage {
    font-size: 2em;
    padding: 1em;
    text-align: center;
    border: 1px solid crimson;
    color: #888;
}

.errorMessage.light {
    background-color: #fff;
    color: #444;
}

.errorMessage.dark {
    background-color: #444;
    color: #fff;
}
