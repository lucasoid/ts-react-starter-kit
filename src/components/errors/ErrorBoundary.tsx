import * as React from 'react';
import { ErrorMessage } from './ErrorMessage';
import { ErrorTypes } from './ErrorTypes';

interface IErrorBoundaryProps {
    children?: any;
}

interface IErrorBoundaryState {
    hasError: boolean;
    errorName: string;
}

export class ErrorBoundary extends React.Component<IErrorBoundaryProps, IErrorBoundaryState> {
    state = {
        hasError: false,
        errorName: '',
    };

    constructor(props) {
        super(props);
    }

    static getDerivedStateFromError(error) {
        let name = ErrorTypes[error.name] || ErrorTypes.Error;
        return { hasError: true, errorName: name };
    }

    componentDidCatch(error, errorInfo) {
        console.error(error, errorInfo);
    }

    render() {
        if (this.state.hasError) {
            switch (this.state.errorName) {
                case ErrorTypes.E404:
                    return <ErrorMessage>⚠ Not found</ErrorMessage>;
                case ErrorTypes.E500:
                case ErrorTypes.Error:
                default:
                    return <ErrorMessage>⚠ Something went wrong</ErrorMessage>;
            }
        } else return this.props.children;
    }
}
