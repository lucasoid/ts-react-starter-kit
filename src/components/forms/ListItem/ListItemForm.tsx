import * as React from 'react';
import uuidv4 from 'uuid/v4';
import { Form } from '~components/forms/Form';
import { TextField } from '~components/forms/Inputs';

interface IListFormProps {
    onSubmit: Function;
    onCancel: Function;
}
export const ListItemForm: React.FC<IListFormProps> = props => {
    const [name, setName] = React.useState('');
    const [category, setCategory] = React.useState('');
    const [quantity, setQuantity] = React.useState(1);

    return (
        <Form
            onSubmit={() => {
                props.onSubmit({
                    id: uuidv4(),
                    name: name,
                    category: category,
                    quantity: quantity,
                    isActive: true,
                });
            }}
            onCancel={props.onCancel}
            submitText={'Create'}
            cancelText={'Cancel'}
        >
            <TextField value={name} label="Name" onChange={setName} />
            <TextField value={category} label="Category" onChange={setCategory} />
            <TextField value={quantity} label="Quantity" onChange={setQuantity} />
        </Form>
    );
};
