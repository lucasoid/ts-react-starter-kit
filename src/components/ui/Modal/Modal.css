.modalWrap {
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: #efefefaa;
}

.modalDialog {
    max-width: 100%;
    margin: 40px auto;
    padding: 1em;
    background: #fff;
    box-sizing: border-box;
}

.modalDialog.small {
    width: 500px;
}

.modalDialog.medium {
    width: 740px;
}

.modalDialog.large {
    width: 1080px;
}

.close {
    text-align: right;
}
