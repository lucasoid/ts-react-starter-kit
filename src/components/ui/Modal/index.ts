export { Modal, ModalWidth } from './Modal';
export const modalRootId: string = 'modal-root';
