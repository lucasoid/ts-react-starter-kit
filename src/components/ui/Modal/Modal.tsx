import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Button } from '~components/ui/Button';
import { modalRootId } from '~components/ui/Modal';
import * as styles from './Modal.css';

export enum ModalWidth {
    SMALL,
    MEDIUM,
    LARGE,
}

interface IModalProps {
    width: ModalWidth;
    onDismiss: Function;
    children?: any;
}

export const Modal: React.FC<IModalProps> = props => {
    let el = document.createElement('div');

    React.useEffect(() => {
        document.getElementById(modalRootId).appendChild(el);
        return () => document.getElementById(modalRootId).removeChild(el);
    });

    React.useEffect(() => {
        const handler: EventListener = (e: KeyboardEvent) => {
            if (e.key === 'Escape') {
                props.onDismiss();
            }
        };
        window.addEventListener('keyup', handler);
        return () => window.removeEventListener('keyup', handler);
    });

    return ReactDOM.createPortal(
        <div className={styles.modalWrap}>
            <div
                role={'dialog'}
                className={`${styles.modalDialog} ${
                    props.width === ModalWidth.SMALL
                        ? styles.small
                        : props.width === ModalWidth.LARGE
                        ? styles.large
                        : styles.medium
                }`}
            >
                <div className={styles.close}>
                    <Button plaintext={true} onClick={props.onDismiss} title={'Close'}>
                        âœ•
                    </Button>
                </div>
                {props.children}
            </div>
        </div>,
        el,
    );
};
