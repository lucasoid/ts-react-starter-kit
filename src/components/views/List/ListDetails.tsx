import * as React from 'react';
import { Button, ButtonTypes } from '~components/ui/Button';
import { Loading, Await } from '~components/Loading';
import { ListItem } from './ListItem';
import { Modal, ModalWidth } from '~components/ui/Modal';
import { ListItemForm } from '~components/forms/ListItem';
import {
    fetchListItems,
    IListItem,
    subscribeToListItems,
    unsubscribeToListItems,
    IList,
    fetchLists,
    editListItem,
    createListItem,
} from '~services/shoppingListApi';
import { E404 } from '~components/errors/ErrorTypes';

interface IListDetailsProps {
    listId: string;
}

export const ListDetails: React.FC<IListDetailsProps> = props => {
    let [list, setList] = React.useState({ id: null, name: null, owner: null, members: [] } as IList);
    let [loading, setLoading] = React.useState(true);
    let [items, setItems] = React.useState([]);
    let [notFound, setNotFound] = React.useState(false);
    let [showCreateDialog, setShowCreateDialog] = React.useState(false);
    React.useEffect(() => {
        const subscriber = function(items) {
            setItems(items);
            setLoading(false);
        };
        fetchLists().then(lists => {
            let list = lists.find(_list => _list.id === props.listId);
            if (!list) {
                setNotFound(true);
            } else {
                setList(list);
                subscribeToListItems(props.listId, subscriber);
                fetchListItems(list.id);
            }
        });
        return () => unsubscribeToListItems(props.listId, subscriber);
    }, [props.listId]);

    function toggle(item: IListItem) {
        item.isActive = !item.isActive;
        // update the API, which will trigger the subscriber when it completes
        editListItem(item);
        // update state optimistically while we wait
        let _items = [...items];
        _items.splice(
            items.findIndex(_item => _item.id === item.id),
            1,
            item,
        );
        setItems(_items);
    }

    function handleCreate(item: IListItem) {
        item.listId = props.listId;
        createListItem(item);
        let _items = [...items];
        _items.push(item);
        setItems(_items);
        setShowCreateDialog(false);
    }

    if (notFound) throw new E404('List not found');

    return (
        <Await
            loaded={!loading}
            renderFallback={() => (
                <>
                    <Loading el={'h2'} width={'25%'} lines={1}></Loading>
                    <Loading el={'aside'} width={'25%'} lines={1}></Loading>
                    <Loading el={'aside'} width={'25%'} lines={1}></Loading>
                    <ListItem item={null} key={1} onClick={() => {}} />
                    <ListItem item={null} key={2} onClick={() => {}} />
                    <ListItem item={null} key={3} onClick={() => {}} />
                </>
            )}
        >
            <h2>{list.name}</h2>
            <aside>Owner: {list.owner}</aside>
            <aside>Members: {list.members.join(', ')}</aside>
            {items.map((item, index) => (
                <ListItem item={item} key={item.id} onClick={toggle} />
            ))}
            <Button onClick={() => setShowCreateDialog(true)} type={ButtonTypes.PRIMARY} styles={{ marginTop: '1em' }}>
                + New item
            </Button>
            {showCreateDialog && (
                <Modal width={ModalWidth.MEDIUM} onDismiss={() => setShowCreateDialog(false)}>
                    <ListItemForm onSubmit={handleCreate} onCancel={() => setShowCreateDialog(false)} />
                </Modal>
            )}
        </Await>
    );
};
