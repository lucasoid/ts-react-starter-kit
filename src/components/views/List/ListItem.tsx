import * as React from 'react';
import { Button } from '~components/ui/Button';
import { Loading } from '~components/Loading';
import * as styles from './List.css';
import { ThemeContext, Themes } from '~theme';
import { IListItem } from '~services/shoppingListApi';

interface IListItemProps {
    item: IListItem;
    onClick: Function;
}
export const ListItem: React.FC<IListItemProps> = props => {
    let { item } = props;
    let { theme } = React.useContext(ThemeContext);

    return !item ? (
        <div className={`${styles.listItem} ${styles.active} ${theme === Themes.DARK ? styles.dark : styles.light}`}>
            <div className={styles.checkbox}>
                <Button plaintext={true} styles={{ outline: 'none' }}>
                    ☐
                </Button>
            </div>
            <div className={styles.name}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
            <div className={styles.category}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
            <div className={styles.quantity}>
                <Loading width={'100%'} lines={1} el={'span'} />
            </div>
        </div>
    ) : (
        <div
            className={`
                ${styles.listItem}
                ${item.isActive ? styles.active : styles.inactive}
                ${theme === Themes.DARK ? styles.dark : styles.light}
            `}
            onClick={() => props.onClick(item)}
        >
            <div className={styles.checkbox}>
                <Button plaintext={true} styles={{ outline: 'none' }}>
                    {item.isActive ? '☐' : '☑'}
                </Button>
            </div>
            <div className={styles.name}>{item.name}</div>
            <div className={styles.category}>{item.category}</div>
            <div className={styles.quantity}>{item.quantity}</div>
        </div>
    );
};
