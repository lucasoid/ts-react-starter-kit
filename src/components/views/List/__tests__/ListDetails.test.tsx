import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import { ListDetails } from '../ListDetails';
import { ErrorBoundary } from '~components/errors/ErrorBoundary';
// imported for mocking
import { Modal } from '~components/ui/Modal';
import { ListItemForm } from '~components/forms/ListItem/ListItemForm';

import {
    fetchListItems,
    IListItem,
    subscribeToListItems,
    unsubscribeToListItems,
    IList,
    fetchLists,
    editListItem,
    createListItem,
} from '~services/shoppingListApi';

jest.mock('~components/ui/Modal', () => ({
    Modal: props => (
        <div>
            <button title="close" onClick={props.onDismiss}>
                X
            </button>
            {props.children}
        </div>
    ),
    ModalWidth: { SMALL: 'small', MEDIUM: 'MEDIUM', LARGE: 'large' },
}));

jest.mock('~components/forms/ListItem/ListItemForm', () => ({
    ListItemForm: props => (
        <form>
            <button
                title={'submit'}
                onClick={() => {
                    // simulating a form submission.
                    // i'm not crazy about this, but there doesn't seem to be another way to access the onSubmit functionality.
                    props.onSubmit({
                        id: 'new',
                        name: 'new item',
                        category: '',
                        quantity: 100,
                        isActive: true,
                    });
                }}
            >
                Submit
            </button>
            <button title="cancel" onClick={props.onCancel}>
                Cancel
            </button>
        </form>
    ),
}));

jest.mock('~services/shoppingListApi', () => ({
    fetchLists: jest.fn(),
    fetchListItems: jest.fn(),
    subscribeToListItems: jest.fn(),
    unsubscribeToListItems: jest.fn(),
    editListItem: jest.fn(),
    createListItem: jest.fn(),
}));

describe('ListDetails', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });

    it('shows a loading UI before data is loaded', () => {
        // @ts-ignore
        fetchLists.mockImplementationOnce(() => new Promise(resolve => null)); // never completes
        let wrapper = mount(<ListDetails listId={'1'} />);
        expect(wrapper).toMatchSnapshot();
    });

    it('shows a list details view once data is loaded', async () => {
        let list: IList = { id: '1', name: 'My Cool List', owner: 'me@domain.com', members: [] };
        let items: IListItem[] = [
            { listId: '1', id: '1', name: 'Eggs - dozen', category: 'dairy', quantity: 1, isActive: true },
            { listId: '1', id: '2', name: 'Milk - gallon', category: 'dairy', quantity: 3, isActive: false },
        ];
        // @ts-ignore
        fetchLists.mockImplementationOnce(() => new Promise(resolve => resolve([list])));
        // @ts-ignore
        subscribeToListItems.mockImplementationOnce((id, cb) => cb(items));

        let wrapper = mount(<ListDetails listId={'1'} />);
        // act() allows the component's state to be updated after the effect.
        // awaiting a Promise ensures that the async code in useEffect is run.
        await act(async () => new Promise(setImmediate));
        // tell Enzyme to re-render after the state was updated.
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });

    it('throws an E404 error if the list is not found', async () => {
        let list: IList = { id: '1', name: 'My Cool List', owner: 'me@domain.com', members: [] };
        // @ts-ignore
        fetchLists.mockImplementationOnce(() => new Promise(resolve => resolve([list])));
        let wrapper = mount(
            <ErrorBoundary>
                <ListDetails listId={'2'} />
            </ErrorBoundary>,
        );
        await act(async () => new Promise(setImmediate));
        wrapper.update();
        expect(wrapper).toMatchSnapshot();
    });

    it('allows for opening and closing the create dialog', async () => {
        let list: IList = { id: '1', name: 'My Cool List', owner: 'me@domain.com', members: [] };
        let items: IListItem[] = [];
        // @ts-ignore
        fetchLists.mockImplementationOnce(() => new Promise(resolve => resolve([list])));
        // @ts-ignore
        subscribeToListItems.mockImplementationOnce((id, cb) => cb(items));
        let wrapper = mount(<ListDetails listId={'1'} />);
        await act(async () => new Promise(setImmediate));
        wrapper.update();
        wrapper
            .find('button')
            .first()
            .simulate('click');
        expect(wrapper).toMatchSnapshot();
        // target the button in the mocked Modal:
        wrapper
            .find('button[title="close"]')
            .first()
            .simulate('click');
        expect(wrapper).toMatchSnapshot();
    });

    it('allows for toggling items between active and inactive', async () => {
        let list: IList = { id: '1', name: 'My Cool List', owner: 'me@domain.com', members: [] };
        let items: IListItem[] = [
            { listId: '1', id: '1', name: 'Eggs - dozen', category: 'dairy', quantity: 1, isActive: true },
            { listId: '1', id: '2', name: 'Milk - gallon', category: 'dairy', quantity: 3, isActive: false },
        ];
        // @ts-ignore
        fetchLists.mockImplementationOnce(() => new Promise(resolve => resolve([list])));
        // @ts-ignore
        subscribeToListItems.mockImplementationOnce((id, cb) => cb(items));
        let wrapper = mount(<ListDetails listId={'1'} />);
        await act(async () => new Promise(setImmediate));
        wrapper.update();

        wrapper
            .find('.listItem')
            .first()
            .simulate('click');
        expect(wrapper).toMatchSnapshot();
        wrapper
            .find('.listItem')
            .first()
            .simulate('click');
        expect(wrapper).toMatchSnapshot();
    });

    it('allows for creating items', async () => {
        let list: IList = { id: '1', name: 'My Cool List', owner: 'me@domain.com', members: [] };
        let items: IListItem[] = [];
        // @ts-ignore
        fetchLists.mockImplementationOnce(() => new Promise(resolve => resolve([list])));
        // @ts-ignore
        subscribeToListItems.mockImplementationOnce((id, cb) => cb(items));
        let wrapper = mount(<ListDetails listId={'1'} />);
        await act(async () => new Promise(setImmediate));
        wrapper.update();
        // open the modal
        wrapper
            .find('button')
            .first()
            .simulate('click');
        // submit the mocked form
        wrapper
            .find('button[title="submit"]')
            .first()
            .simulate('click');
        await act(async () => new Promise(setImmediate));
        wrapper.update();
        // the component should now show one item in the list
        expect(wrapper).toMatchSnapshot();
        expect(createListItem).toBeCalled();
    });

    it('allows for dismissing the create form', async () => {
        let list: IList = { id: '1', name: 'My Cool List', owner: 'me@domain.com', members: [] };
        let items: IListItem[] = [];
        // @ts-ignore
        fetchLists.mockImplementationOnce(() => new Promise(resolve => resolve([list])));
        // @ts-ignore
        subscribeToListItems.mockImplementationOnce((id, cb) => cb(items));
        let wrapper = mount(<ListDetails listId={'1'} />);
        await act(async () => new Promise(setImmediate));
        wrapper.update();
        // open the modal
        wrapper
            .find('button')
            .first()
            .simulate('click');
        // cancel the mocked form
        wrapper
            .find('button[title="cancel"]')
            .first()
            .simulate('click');
        expect(wrapper).toMatchSnapshot();
    });
});
