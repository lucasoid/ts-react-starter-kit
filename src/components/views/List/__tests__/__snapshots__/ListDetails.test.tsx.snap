// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListDetails allows for creating items 1`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={true}
    renderFallback={[Function]}
  >
    <h2>
      My Cool List
    </h2>
    <aside>
      Owner: 
      me@domain.com
    </aside>
    <aside>
      Members: 
    </aside>
    <ListItem
      item={
        Object {
          "category": "",
          "id": "new",
          "isActive": true,
          "listId": "1",
          "name": "new item",
          "quantity": 100,
        }
      }
      key="new"
      onClick={[Function]}
    >
      <div
        className="
                listItem
                active
                light
            "
        onClick={[Function]}
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☐
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          new item
        </div>
        <div
          className="category"
        />
        <div
          className="quantity"
        >
          100
        </div>
      </div>
    </ListItem>
    <Button
      onClick={[Function]}
      styles={
        Object {
          "marginTop": "1em",
        }
      }
      type={0}
    >
      <button
        className="btn primary"
        onClick={[Function]}
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        + New item
      </button>
    </Button>
  </Await>
</ListDetails>
`;

exports[`ListDetails allows for dismissing the create form 1`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={true}
    renderFallback={[Function]}
  >
    <h2>
      My Cool List
    </h2>
    <aside>
      Owner: 
      me@domain.com
    </aside>
    <aside>
      Members: 
    </aside>
    <Button
      onClick={[Function]}
      styles={
        Object {
          "marginTop": "1em",
        }
      }
      type={0}
    >
      <button
        className="btn primary"
        onClick={[Function]}
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        + New item
      </button>
    </Button>
  </Await>
</ListDetails>
`;

exports[`ListDetails allows for opening and closing the create dialog 1`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={true}
    renderFallback={[Function]}
  >
    <h2>
      My Cool List
    </h2>
    <aside>
      Owner: 
      me@domain.com
    </aside>
    <aside>
      Members: 
    </aside>
    <Button
      onClick={[Function]}
      styles={
        Object {
          "marginTop": "1em",
        }
      }
      type={0}
    >
      <button
        className="btn primary"
        onClick={[Function]}
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        + New item
      </button>
    </Button>
    <Modal
      onDismiss={[Function]}
      width="MEDIUM"
    >
      <div>
        <button
          onClick={[Function]}
          title="close"
        >
          X
        </button>
        <ListItemForm
          onCancel={[Function]}
          onSubmit={[Function]}
        >
          <form>
            <button
              onClick={[Function]}
              title="submit"
            >
              Submit
            </button>
            <button
              onClick={[Function]}
              title="cancel"
            >
              Cancel
            </button>
          </form>
        </ListItemForm>
      </div>
    </Modal>
  </Await>
</ListDetails>
`;

exports[`ListDetails allows for opening and closing the create dialog 2`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={true}
    renderFallback={[Function]}
  >
    <h2>
      My Cool List
    </h2>
    <aside>
      Owner: 
      me@domain.com
    </aside>
    <aside>
      Members: 
    </aside>
    <Button
      onClick={[Function]}
      styles={
        Object {
          "marginTop": "1em",
        }
      }
      type={0}
    >
      <button
        className="btn primary"
        onClick={[Function]}
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        + New item
      </button>
    </Button>
  </Await>
</ListDetails>
`;

exports[`ListDetails allows for toggling items between active and inactive 1`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={true}
    renderFallback={[Function]}
  >
    <h2>
      My Cool List
    </h2>
    <aside>
      Owner: 
      me@domain.com
    </aside>
    <aside>
      Members: 
    </aside>
    <ListItem
      item={
        Object {
          "category": "dairy",
          "id": "1",
          "isActive": false,
          "listId": "1",
          "name": "Eggs - dozen",
          "quantity": 1,
        }
      }
      key="1"
      onClick={[Function]}
    >
      <div
        className="
                listItem
                inactive
                light
            "
        onClick={[Function]}
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☑
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          Eggs - dozen
        </div>
        <div
          className="category"
        >
          dairy
        </div>
        <div
          className="quantity"
        >
          1
        </div>
      </div>
    </ListItem>
    <ListItem
      item={
        Object {
          "category": "dairy",
          "id": "2",
          "isActive": false,
          "listId": "1",
          "name": "Milk - gallon",
          "quantity": 3,
        }
      }
      key="2"
      onClick={[Function]}
    >
      <div
        className="
                listItem
                inactive
                light
            "
        onClick={[Function]}
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☑
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          Milk - gallon
        </div>
        <div
          className="category"
        >
          dairy
        </div>
        <div
          className="quantity"
        >
          3
        </div>
      </div>
    </ListItem>
    <Button
      onClick={[Function]}
      styles={
        Object {
          "marginTop": "1em",
        }
      }
      type={0}
    >
      <button
        className="btn primary"
        onClick={[Function]}
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        + New item
      </button>
    </Button>
  </Await>
</ListDetails>
`;

exports[`ListDetails allows for toggling items between active and inactive 2`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={true}
    renderFallback={[Function]}
  >
    <h2>
      My Cool List
    </h2>
    <aside>
      Owner: 
      me@domain.com
    </aside>
    <aside>
      Members: 
    </aside>
    <ListItem
      item={
        Object {
          "category": "dairy",
          "id": "1",
          "isActive": true,
          "listId": "1",
          "name": "Eggs - dozen",
          "quantity": 1,
        }
      }
      key="1"
      onClick={[Function]}
    >
      <div
        className="
                listItem
                active
                light
            "
        onClick={[Function]}
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☐
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          Eggs - dozen
        </div>
        <div
          className="category"
        >
          dairy
        </div>
        <div
          className="quantity"
        >
          1
        </div>
      </div>
    </ListItem>
    <ListItem
      item={
        Object {
          "category": "dairy",
          "id": "2",
          "isActive": false,
          "listId": "1",
          "name": "Milk - gallon",
          "quantity": 3,
        }
      }
      key="2"
      onClick={[Function]}
    >
      <div
        className="
                listItem
                inactive
                light
            "
        onClick={[Function]}
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☑
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          Milk - gallon
        </div>
        <div
          className="category"
        >
          dairy
        </div>
        <div
          className="quantity"
        >
          3
        </div>
      </div>
    </ListItem>
    <Button
      onClick={[Function]}
      styles={
        Object {
          "marginTop": "1em",
        }
      }
      type={0}
    >
      <button
        className="btn primary"
        onClick={[Function]}
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        + New item
      </button>
    </Button>
  </Await>
</ListDetails>
`;

exports[`ListDetails shows a list details view once data is loaded 1`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={true}
    renderFallback={[Function]}
  >
    <h2>
      My Cool List
    </h2>
    <aside>
      Owner: 
      me@domain.com
    </aside>
    <aside>
      Members: 
    </aside>
    <ListItem
      item={
        Object {
          "category": "dairy",
          "id": "1",
          "isActive": true,
          "listId": "1",
          "name": "Eggs - dozen",
          "quantity": 1,
        }
      }
      key="1"
      onClick={[Function]}
    >
      <div
        className="
                listItem
                active
                light
            "
        onClick={[Function]}
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☐
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          Eggs - dozen
        </div>
        <div
          className="category"
        >
          dairy
        </div>
        <div
          className="quantity"
        >
          1
        </div>
      </div>
    </ListItem>
    <ListItem
      item={
        Object {
          "category": "dairy",
          "id": "2",
          "isActive": false,
          "listId": "1",
          "name": "Milk - gallon",
          "quantity": 3,
        }
      }
      key="2"
      onClick={[Function]}
    >
      <div
        className="
                listItem
                inactive
                light
            "
        onClick={[Function]}
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☑
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          Milk - gallon
        </div>
        <div
          className="category"
        >
          dairy
        </div>
        <div
          className="quantity"
        >
          3
        </div>
      </div>
    </ListItem>
    <Button
      onClick={[Function]}
      styles={
        Object {
          "marginTop": "1em",
        }
      }
      type={0}
    >
      <button
        className="btn primary"
        onClick={[Function]}
        style={
          Object {
            "marginTop": "1em",
          }
        }
      >
        + New item
      </button>
    </Button>
  </Await>
</ListDetails>
`;

exports[`ListDetails shows a loading UI before data is loaded 1`] = `
<ListDetails
  listId="1"
>
  <Await
    loaded={false}
    renderFallback={[Function]}
  >
    <Loading
      el="h2"
      lines={1}
      width="25%"
    >
      <h2
        key="0"
        style={
          Object {
            "width": "25%",
          }
        }
      >
        <span
          className="loadingText light"
        />
      </h2>
    </Loading>
    <Loading
      el="aside"
      lines={1}
      width="25%"
    >
      <aside
        key="0"
        style={
          Object {
            "width": "25%",
          }
        }
      >
        <span
          className="loadingText light"
        />
      </aside>
    </Loading>
    <Loading
      el="aside"
      lines={1}
      width="25%"
    >
      <aside
        key="0"
        style={
          Object {
            "width": "25%",
          }
        }
      >
        <span
          className="loadingText light"
        />
      </aside>
    </Loading>
    <ListItem
      item={null}
      key="1"
      onClick={[Function]}
    >
      <div
        className="listItem active light"
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☐
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
        <div
          className="category"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
        <div
          className="quantity"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
      </div>
    </ListItem>
    <ListItem
      item={null}
      key="2"
      onClick={[Function]}
    >
      <div
        className="listItem active light"
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☐
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
        <div
          className="category"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
        <div
          className="quantity"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
      </div>
    </ListItem>
    <ListItem
      item={null}
      key="3"
      onClick={[Function]}
    >
      <div
        className="listItem active light"
      >
        <div
          className="checkbox"
        >
          <Button
            plaintext={true}
            styles={
              Object {
                "outline": "none",
              }
            }
          >
            <button
              className="btnPlaintext secondary"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                }
              }
            >
              ☐
            </button>
          </Button>
        </div>
        <div
          className="name"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
        <div
          className="category"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
        <div
          className="quantity"
        >
          <Loading
            el="span"
            lines={1}
            width="100%"
          >
            <span
              key="0"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <span
                className="loadingText light"
              />
            </span>
          </Loading>
        </div>
      </div>
    </ListItem>
  </Await>
</ListDetails>
`;

exports[`ListDetails throws an E404 error if the list is not found 1`] = `
<ErrorBoundary>
  <ErrorMessage>
    <div
      className="errorMessage light"
    >
      ⚠ Not found
    </div>
  </ErrorMessage>
</ErrorBoundary>
`;
